mmv1/third_party/terraform/services/kms/data_source_google_kms_crypto_key_versions.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		return err
+	}
+
+	cryptoKeyId, err := ParseKmsCryptoKeyId(d.Get("crypto_key").(string), config)
+	if err != nil {
+		return err
+	}
+
+	id := fmt.Sprintf("%s/cryptoKeyVersions", cryptoKeyId.CryptoKeyId())
+	d.SetId(id)
+
+	log.Printf("[DEBUG] Searching for cryptoKeyVersions in crypto key %s", cryptoKeyId.CryptoKeyId())
+	versions, err := dataSourceKMSCryptoKeyVersionsList(d, meta, cryptoKeyId.CryptoKeyId(), userAgent)
+	if err != nil {
+		return err
+	}
+
+	if len(versions) > 0 {
+		log.Printf("[DEBUG] Found %d cryptoKeyVersions in crypto key %s", len(versions), cryptoKeyId.CryptoKeyId())
+		value, err := flattenKMSCryptoKeyVersionsList(d, config, versions, cryptoKeyId.CryptoKeyId())
+		if err != nil {
+			return fmt.Errorf("error flattening cryptoKeyVersions list: %s", err)
+		}
+		if err := d.Set("versions", value); err != nil {
+			return fmt.Errorf("error setting versions: %s", err)
+		}
+	} else {
+		log.Printf("[DEBUG] Found 0 versions in crypto key %s", cryptoKeyId.CryptoKeyId())
+	}
+
+	url, err := tpgresource.ReplaceVars(d, config, "{{KMSBasePath}}{{crypto_key}}")
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Getting purpose of CryptoKey: %#v", url)
+	res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+		Config:    config,
+		Method:    "GET",
+		Project:   cryptoKeyId.KeyRingId.Project,
+		RawURL:    url,
+		UserAgent: userAgent,
+	})
+	if err != nil {
+		return transport_tpg.HandleDataSourceNotFoundError(err, d, fmt.Sprintf("KmsCryptoKey %q", d.Id()), url)
+	}
+
+	if res["purpose"] == "ASYMMETRIC_SIGN" || res["purpose"] == "ASYMMETRIC_DECRYPT" {
+		url, err = tpgresource.ReplaceVars(d, config, fmt.Sprintf("{{KMSBasePath}}{{crypto_key}}/cryptoKeyVersions/%d/publicKey", d.Get("versions.0.version")))
+		if err != nil {
+			return err
+		}
+		log.Printf("[DEBUG] Getting public key of CryptoKeyVersion: %#v", url)
+
+		res, err = transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+			Config:               config,
+			Method:               "GET",
+			Project:              cryptoKeyId.KeyRingId.Project,
+			RawURL:               url,
+			UserAgent:            userAgent,
+			Timeout:              d.Timeout(schema.TimeoutRead),
+			ErrorRetryPredicates: []transport_tpg.RetryErrorPredicateFunc{transport_tpg.IsCryptoKeyVersionsPendingGeneration},
+		})
+
+		if err != nil {
+			log.Printf("Error generating public key: %s", err)
+			return err
+		}
+
+		if err := d.Set("public_key", flattenKmsCryptoKeyVersionPublicKey(res, d)); err != nil {
+			return fmt.Errorf("Error setting CryptoKeyVersion public key: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func dataSourceKMSCryptoKeyVersionsList(d *schema.ResourceData, meta interface{}, cryptoKeyId string, userAgent string) ([]interface{}, error) {
+	config := meta.(*transport_tpg.Config)
+
+	url, err := tpgresource.ReplaceVars(d, config, "{{KMSBasePath}}{{crypto_key}}/cryptoKeyVersions?filter=state=ENABLED")
```

Comment:
Do we want to default the filter to only ENABLED keys? We should at least document that if it's what we want

I'm worried that a user would try to get *all* their keys and be confused when only enabled ones come back

Reply:
Is this the correct format for a filter also? filter=state=enabled seems like it should be escaped somehow

Reply:
> Do we want to default the filter to only ENABLED keys? We should at least document that if it's what we want
> 
> I'm worried that a user would try to get _all_ their keys and be confused when only enabled ones come back

I think to avoid confusion we can have it include both enabled and disabled versions. My view point is that users would only be interested in ENABLED versions since you are unable to do anything with disable versions until they are enabled.

The format is also correct, i agree it's a bit weird with the current format of `filter=state=enabled`

Reply:
latest commit includes the filter so that users can filter through the list of versions (such as name, state, etc.)